Move pylama to tox

vitorbaptista opened the pull request:
Tox is the go-to tool for our testing process, including linting. By moving
pylama to it, the only development dependency needed is `tox`, it'll take care
of creating the separate environments. Considering that the user already have
`tox` installed, the only thing she needs to do to run our entire test
suite (which includes linting) is:

```
tox
```

vitorbaptista committed:
Move pylama to tox

Tox is the go-to tool for our testing process, including linting. By moving
pylama to it, the only development dependency needed is `tox`, it'll take care
of creating the separate environments. Considering that the user already have
`tox` installed, the only thing she needs to do to run our entire test
suite (which includes linting) is:

```
tox
```

roll commented:
@pwalsh 
@vitorbaptista 
Doesn't this change break linting on Travis?

I'm going to be pretty honest here. It's a very small change but in my mind it's related to the important high-level approach. So TBH I'm against this change.

This change mostly affects the Frictionles Data libraries which is under my responsibility. If we consider this change only for new repositories it's one discussion. But because this change is already applied for `goodtables-py` it means that it's not this case.

In the last two years I've established and polished almost every process going with this libs to reduce the maintenance cost. E.g. there are 10+ Python repositories that's 100% synced making it much easier to work with all of them. The process is time-proven and battle tested. It's far from perfect. But it's guaranteed to work.

An important engineering principle is to do not break working things. If we going to change something already proven to work we should understand what's additional value we bring. Is this value is more important than cost/risk of changes etc.

**I really don't care about how we run linting** etc. My point is that a week ago I was responsible for 15 repositories synced/proven to work. And now I still responsible for 14 repositories synced/proven to work and `goodtables-py` where I'm not even sure linting works at all. And I don't see how changes like this help us to deliver e.g. for ODI project.

Disclaimer: this comment is more related to the Frictionless Data repos. If this change works I'm neutral on changing it for `oki-py`. But still a question - probably there should be a big win of this change to make all existent OKI repos unsync with this reference repo.

vitorbaptista commented:
@roll

It doesn't break linting, the only thing that changes is that `tox` handles all testing dependencies. The build is currently broken because pytest dropped support for python 3.3, and there's a bug on setuptools on 3.4 (https://github.com/pypa/setuptools/issues/951)

Tox is *the* test runner for Python, and as lint is a type of tests, it should be handled by tox itself IMHO. Not only tests, but also things like building the sphinx documentation. For examples, check https://github.com/django/django/blob/master/tox.ini, https://github.com/pallets/flask/blob/master/tox.ini, https://github.com/zzzeek/sqlalchemy/blob/master/tox.ini.

I agree that we shouldn't change things that are working. However, this is a very small change that doesn't add any extra dependencies, just improves the usage of our current tools.

roll commented:
@vitorbaptista 
If I remember correctly here in Travis we use a matrix of the env variables to run one `tox` environment per one `Travis` executable:

```yml
python:
  - 2.7
  - 3.3
  - 3.4
  - 3.5

env:
  global:
    - TOXENV="py${PYTHON_VERSION//./}"

script:
  - make test
```

So if we make `lint` to be a `tox` environment we need to change the way how we run `tox` in Travis. I could be wrong here because I almost never use `tox` in my work because it too slow. So I don't know how it works in details. But I can't figure out where we call `pylama` now in our builds - https://travis-ci.org/frictionlessdata/goodtables-py/jobs/310054598. So just not sure TBH.

pwalsh commented:
@roll 

I'm really having trouble following your line of thinking here. This is not about you, or the Frictionless Data libraries, or breaking a process that worked over the last two years. When I formalised our coding standards and, this repo for our team, I also did so in a way that was incompatible with prior art at OKI. Things move forward and it is totally ok.

Here, we are talking about a trivial change, but an important one for simplification. This change is one that I have supported since @vitorbaptista gave his talk at PyConUK, we [discussed it here](https://github.com/okfn/coding-standards/issues/4), and now it is great that the time has come to do a PR and add it as our team best practice.

This does not mean we need to go and change the 300+ code repositories that OKI has, nor the ones that you maintain, in the same way that when we introduced this process, we did not go and change all existing code bases to fit it.

roll commented:
I pretty agree on continuous evolution of our approaches. Including packaging. And again I think this change is absolutely irrelevant - an every working option should be OK for everyone. It really doesn't matter in high-level.

My point is that first we're better to discuss, test and agree and only then to incorporate changes into working projects.

vitorbaptista committed:
Enable Python 3.6, remove 3.3 and disable 3.4

Python 3.3 isn't supported anymore, and there's a bug with setuptools on 3.4
that breaks our build.

vitorbaptista commented:
@roll I agree with you, we need to discuss those changes before implementing them. But isn't this the whole point of doing PRs? The code isn't complete when it comes to the PR, I'm happy to change anything, or even drop it. For instance, you were right: linting was not running on Travis. Thanks to your code review, I fixed it and it's now running (https://travis-ci.org/okfn/oki-py/jobs/319654868).

For bigger changes, I agree that we should create an issue and talk about it before actually doing it. This is something I should get better at. We just need to find the balance, as it's easy to loose time [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding).

Are you OK with these changes, now that linting is fixed?

roll commented:
@vitorbaptista 
Yes of course - why not.

I believe in separation of concerns and in that people should be able to implement their vision on their segments of work. Improving our packaging approach is clearly an Engineering Lead responsibility. So even I'm not totally sold on this exact change I'm just trust your vision on it. Bug has been fixed so it's good to go I think.

> We just need to find the balance, as it's easy to loose time bikeshedding.

That's exactly why I so value an approach of having explicit rules on decision making, responsibility separation etc. It saves a lot of time and makes things much easier for everyone.

---

Disclaimer: we're not crazy - this discussion wasn't really about linting :smiley:
